README.rst
Python Client for Grafeas API
=============================

|stable| |pypi| |versions|

`Grafeas API`_: An implementation of the Grafeas API, which stores, and enables querying and retrieval of critical metadata about all of your software artifacts.

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/grafeas.svg
   :target: https://pypi.org/project/grafeas/
.. |versions| image:: https://img.shields.io/pypi/pyversions/grafeas.svg
   :target: https://pypi.org/project/grafeas/
.. _Grafeas API: https://grafeas.io
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/containeranalysis/latest
.. _Product Documentation:  https://grafeas.io

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Grafeas API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Grafeas API.:  https://grafeas.io
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install grafeas


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install grafeas

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Grafeas API
   to see other available methods on the client.
-  Read the `Grafeas API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Grafeas API Product documentation:  https://grafeas.io
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

[PyPI History][1]

[1]: https://pypi.org/project/grafeas/#history

## [1.9.0](https://github.com/googleapis/python-grafeas/compare/v1.8.1...v1.9.0) (2023-10-20)


### Features

* Add enum for grafeas v1 cvss to support both cvss v2 and v3 ([#241](https://github.com/googleapis/python-grafeas/issues/241)) ([a44411d](https://github.com/googleapis/python-grafeas/commit/a44411d7719d4870da7b8444652abb515ab13268))
* Add VULNERABILITY_ASSESSMENT Note type ([16848a7](https://github.com/googleapis/python-grafeas/commit/16848a7041efd25ddb209e242e6bf3af3a5c0936))
* Adds Vex_Assessment derived from the Note to resources' occurrences ([16848a7](https://github.com/googleapis/python-grafeas/commit/16848a7041efd25ddb209e242e6bf3af3a5c0936))
* Enable "rest" transport in Python for services supporting numeric enums ([c1c2e7c](https://github.com/googleapis/python-grafeas/commit/c1c2e7c8ef77d4b4d4c8452feb7ae4b8055957ed))
* Import of Grafeas from Github ([c1c2e7c](https://github.com/googleapis/python-grafeas/commit/c1c2e7c8ef77d4b4d4c8452feb7ae4b8055957ed))
* VEX notes can now be written to add CVE assessments ([16848a7](https://github.com/googleapis/python-grafeas/commit/16848a7041efd25ddb209e242e6bf3af3a5c0936))


### Bug Fixes

* Add async context manager return types ([#248](https://github.com/googleapis/python-grafeas/issues/248)) ([34fc787](https://github.com/googleapis/python-grafeas/commit/34fc78784aab13b12a1012c9881485ba9956c4c5))


### Documentation

* Fix formatting of request arg in docstring ([#240](https://github.com/googleapis/python-grafeas/issues/240)) ([05a0571](https://github.com/googleapis/python-grafeas/commit/05a0571e39691e677a53ceda7eae7ac915524cba))
* Minor formatting ([#260](https://github.com/googleapis/python-grafeas/issues/260)) ([8eebac7](https://github.com/googleapis/python-grafeas/commit/8eebac71084cb8339155484f7b269b5c784cea2d))

## [1.8.1](https://github.com/googleapis/python-grafeas/compare/v1.8.0...v1.8.1) (2023-01-20)


### Bug Fixes

* Add context manager return types ([2b37951](https://github.com/googleapis/python-grafeas/commit/2b3795107feb1832670f3db11418695258d65c53))


### Documentation

* Add documentation for enums ([2b37951](https://github.com/googleapis/python-grafeas/commit/2b3795107feb1832670f3db11418695258d65c53))

## [1.8.0](https://github.com/googleapis/python-grafeas/compare/v1.7.0...v1.8.0) (2023-01-10)


### Features

* Add support for python 3.11 ([#226](https://github.com/googleapis/python-grafeas/issues/226)) ([c13c260](https://github.com/googleapis/python-grafeas/commit/c13c2602ea0b36084123114845ea03b377a292df))

## [1.7.0](https://github.com/googleapis/python-grafeas/compare/v1.6.1...v1.7.0) (2022-12-14)


### Features

* Add support for `grafeas.grafeas.__version__` ([ad9ec08](https://github.com/googleapis/python-grafeas/commit/ad9ec0886ebf5be225f0ac91fd2bb0c45ab6ad7b))
* Add typing to proto.Message based class attributes ([ad9ec08](https://github.com/googleapis/python-grafeas/commit/ad9ec0886ebf5be225f0ac91fd2bb0c45ab6ad7b))


### Bug Fixes

* Add dict typing for client_options ([ad9ec08](https://github.com/googleapis/python-grafeas/commit/ad9ec0886ebf5be225f0ac91fd2bb0c45ab6ad7b))
* **deps:** Require google-api-core &gt;=1.34.0, >=2.11.0  ([ad9ec08](https://github.com/googleapis/python-grafeas/commit/ad9ec0886ebf5be225f0ac91fd2bb0c45ab6ad7b))
* Drop usage of pkg_resources ([ad9ec08](https://github.com/googleapis/python-grafeas/commit/ad9ec0886ebf5be225f0ac91fd2bb0c45ab6ad7b))
* Fix timeout default values ([ad9ec08](https://github.com/googleapis/python-grafeas/commit/ad9ec0886ebf5be225f0ac91fd2bb0c45ab6ad7b))


### Documentation

* **samples:** Snippetgen handling of repeated enum field ([ad9ec08](https://github.com/googleapis/python-grafeas/commit/ad9ec0886ebf5be225f0ac91fd2bb0c45ab6ad7b))
* **samples:** Snippetgen should call await on the operation coroutine before calling result ([ad9ec08](https://github.com/googleapis/python-grafeas/commit/ad9ec0886ebf5be225f0ac91fd2bb0c45ab6ad7b))

## [1.6.1](https://github.com/googleapis/python-grafeas/compare/v1.6.0...v1.6.1) (2022-10-07)


### Bug Fixes

* **deps:** Allow protobuf 3.19.5 ([#216](https://github.com/googleapis/python-grafeas/issues/216)) ([49cf6f2](https://github.com/googleapis/python-grafeas/commit/49cf6f2509d2750cca42d75176c632afd46a2d8e))

## [1.6.0](https://github.com/googleapis/python-grafeas/compare/v1.5.1...v1.6.0) (2022-10-03)


### Features

* Add new analysis status and cvss version fields ([#214](https://github.com/googleapis/python-grafeas/issues/214)) ([f3aaadb](https://github.com/googleapis/python-grafeas/commit/f3aaadb44d99610b056565169d728d762c0d0465))


### Bug Fixes

* **deps:** Require protobuf >= 3.20.2 ([#213](https://github.com/googleapis/python-grafeas/issues/213)) ([6a28e86](https://github.com/googleapis/python-grafeas/commit/6a28e86042bd3ad7eb7c9096d7e6486cd40878c8))

## [1.5.1](https://github.com/googleapis/python-grafeas/compare/v1.5.0...v1.5.1) (2022-08-11)


### Bug Fixes

* **deps:** allow protobuf < 5.0.0 ([#198](https://github.com/googleapis/python-grafeas/issues/198)) ([e66561a](https://github.com/googleapis/python-grafeas/commit/e66561a9ed210a0564b5027a1decb7f36413e447))
* **deps:** require proto-plus >= 1.22.0 ([e66561a](https://github.com/googleapis/python-grafeas/commit/e66561a9ed210a0564b5027a1decb7f36413e447))

## [1.5.0](https://github.com/googleapis/python-grafeas/compare/v1.4.5...v1.5.0) (2022-07-14)


### Features

* Add `Digest`, `FileLocation` and `License` ([#186](https://github.com/googleapis/python-grafeas/issues/186)) ([69b5e8b](https://github.com/googleapis/python-grafeas/commit/69b5e8b7fe9162f93a2141a30a94e0bf637af433))
* add audience parameter ([59cb75d](https://github.com/googleapis/python-grafeas/commit/59cb75d9b2beeaaa25ab4cd057bcbd5efc5796d7))


### Bug Fixes


owl-bot-staging/v1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.7",
    "3.8",
    "3.9",
    "3.10",
    "3.11",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==22.3.0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.11"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=grafeas/grafeas_v1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install(
        'mypy',
        'types-requests',
        'types-protobuf'
    )
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'grafeas',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
